IZ1 C/C++, RunID 1208

IMadiev opened the pull request.

IMadiev committed:
IZ1 C/C++

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+*/
+
+// Решение
+#include "stdio.h"
```
стандартные библиотеки нужно подключать через <>

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+// Функция поиска нужных строк
+char **filter (char **m, size_t size)
+{
+	char str1[] = "From:";
```
лучше было бы поместить все строки в отдельный массив строк

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+	char **result=NULL;
+	size_t str_num=0;
+	result=(char**)realloc(result, (str_num+1)*sizeof(char*));
+	if (result==NULL) {printf ("[error]"); return 0;}
```
если realloc возвращает NULL, то он не очищает старую память, поэтому её нужно очищать самостоятельно
тут же Вы затираете старый указатель, поэтому в случае ошибки у Вас будет утечка памяти

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+		{
+			result[str_num]=m[i];
+			str_num++;
+			result=(char**)realloc(result, (str_num+1)*sizeof(char*));
```
realloc на каждой вставке - плохая практика, приводящая к квадратичной вычислительной сложности вместо амортизированной константной

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+	size_t size_file = 0;
+	char **file = (char**)malloc(sizeof(char*));
+	if (file == NULL) { printf("[error]"); return 0; }
+	char* buffer = (char*)malloc(2048*sizeof(char));
```
что такое 2048?

leshiy1295 replied:
и зачем нужно память под такой буфер выделять динамически?

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+	char **file = (char**)malloc(sizeof(char*));
+	if (file == NULL) { printf("[error]"); return 0; }
+	char* buffer = (char*)malloc(2048*sizeof(char));
+	if (buffer == NULL) { printf("[error]"); return 0; }
```
в случае ошибки будет утечка памяти под file

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+	while (fgets(buffer, 2048, stdin) != NULL)
+	{
+		size_t size_str = 0;
+		while (buffer[size_str] != '\0') 
```
всегда стоит ставить фигурные скобки с целью избежания глупых ошибок в будущем

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+}
+
+
+int main()
```
огромная функция main
нужно декомпозировать на отдельные функции
оформить имеет смысл в следующем виде:
- функция ввода и инициализации данных
- обработка
- функция вывода результатов и освобождения выделенных ресурсов
при этом имеет смысл сделать разделение на интерфейс и реализацию и представить программный файл в виде:
- описания функций
- определение main
- определения всех остальных функций

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+		while (buffer[size_str] != '\0') 
+			size_str++;
+		size_str++;
+		file[size_file] = (char*)malloc(size_str * sizeof(char));
```
что, если произошла ошибка?

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+			file[size_file][i] = buffer[i];
+
+		size_file++;
+		file = (char**)realloc(file, (size_file + 1) * sizeof(char*));
```
ещё одна утечка

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+	}
+	//освобождение буфера
+	free(buffer);
+	//проверка на пустой массив
```
из кода и так ясен код - не нужно дублировать подобными комментариями

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+	//проверка на пустой массив
+	if (size_file == 0) { free(file); return 0; }
+	//вывод массива
+	/*for (size_t i = 0; i < size_file; ++i) 
```
нерабочего закомментиованного кода быть не должно

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,102 @@
+		size_t size_str = 0;
+		while (buffer[size_str] != '\0') 
+			size_str++;
+		size_str++;
```
неочевидно, зачем нужно делать дополнительно ++ - вот подобные места можно снабжать комментариями или рефакторить, чтобы вопросов не возникало

leshiy1295 reviewed (changes requested):
постарался во всех местах продублировать ошибки
в дальнейшем буду оставлять замечания на первую встреченную ошибку - подобные замечания нужно будет вам обнаружить и исправить самостоятельно

IMadiev committed:
Update IZ-1 - RunID1564

1. изменены кавычки стандартных библиотек
2. доработаны обработки ошибок - убраны ситуации "если realloc возвращает NULL, то он не очищает старую память"
3. убран realloc на каждой вставке в фильтре. 
4. добавлены комментарии
5. упрощена функция main. сложные операции вынесены в отдельные функции
6.  искомые подстроки собраны в массив

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,160 @@
+		while (file[size_file] != NULL) 
+			size_file++;
+	// Удаление ненужного NULL в конце
+	file=(char**)realloc(file, (size_file*sizeof(char*)));
```
что если realloc упадёт?

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,160 @@
+int main()
+{
+	char **file=input_mail();
+	if (file==NULL) return 0;
```
стоит предпочитать полноценный блок с фигурными скобками

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,160 @@
+	file=(char**)realloc(file, (size_file*sizeof(char*)));
+	
+	char **result=filter(file, size_file);
+	if (result==NULL) return 0;
```
в таком случае будет утечка памяти по file

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,160 @@
+	output_result(result);
+
+	// очищение массива
+	for (size_t j = 0; j < size_file; ++j) {
```
это можно было выделить в отдельную функцию
file выделялся не тут, поэтому чисто семантически main не должен знать о структуре file

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,160 @@
+	size_t size_file = 0;
+	char **file = (char**)malloc(sizeof(char*));
+	if (file == NULL) { printf("[error]"); return NULL; }
+	char buffer[2048];
```
все переменные стоит явно инициализировать

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,160 @@
+	if (file == NULL) { printf("[error]"); return NULL; }
+	char buffer[2048];
+	//чтение файла
+	while (fgets(buffer, 2048, stdin) != NULL)
```
одна и та же константа повторяется в нескольких местах - можно её было вынести в define

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,160 @@
+				free(file[j]);
+			}
+			free(file);
+			return 0; 
```
возвращается не число, а указатель - это стоит явно показывать с помощью return NULL

leshiy1295 commented on C-C++/IZ-1.c:
```diff
@@ -0,0 +1,160 @@
+	result=(char**)malloc((buf)*sizeof(char*));
+	if (result==NULL) {printf ("[error]"); return NULL;}
+	for (size_t i=0; i<size; i++)
+		if ((strstr(m[i], str[0])==m[i])||(strstr(m[i], str[1])==m[i])||(strstr(m[i], str[2])==m[i])||(strstr(m[i], str[3])==m[i]))
```
что если в функцию передадут m == NULL?
